### PR TITLE
typescript refactor

### PR DESCRIPTION
**Pull request by atomicpages**
## Summary

* Swap out code for ts
* Generate flow typings so existing code doesn't break
* Run prettier on README

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,46 +1,52 @@
...
+    "description": "",
+    "keywords": [
+        "pagination",
+        "paging",
+        "pager",
+        "table"
+    ],
+    "sideEffects": false,
+    "scripts": {
+        "build": "pika build"
```
Comment (erictooth) [inline package.json]: The change to `pika build` causes the build to break. I know it's the recommendation from @pika/pack, but I'm not sure why it doesn't work.

File: package.json [hunk 2]
```diff
@@ -1,46 +1,52 @@
...
+            [
+                "@pika/plugin-build-web"
+            ],
+            [
+                "pika-plugin-typedefs-to-flow"
+            ]
+        ]
+    },
+    "peerDependencies": {
+        "react": "^16.8.0 || ^17.0.0"
```
Comment (erictooth) [inline package.json]: I've started using `>=16.8.0` in my other libraries; what do you think should be preferred here?

File: src/index.ts [hunk 3]
```diff
@@ -38,20 +39,23 @@ export const getPaginationState = ({
     };
 };
 
-type CurrentPageReducerActions = {| type: "SET", page: number |} | {| type: "NEXT" | "PREV" |};
+type CurrentPageReducerActions = React.Reducer<
+    number,
+    Partial<{ type: "SET"; page: number } | { type: "NEXT" | "PREV" }>
```
Comment (erictooth) [inline src/index.ts]: I'm worried the `Partial` here will cause all fields to be optional, when really we want to preserve this as a discriminated union type

File: src/index.ts [hunk 4]
```diff
@@ -79,41 +83,35 @@ export function usePagination({
...
-                });
-            },
-            [dispatch]
-        ),
+        setPage: React.useCallback((page: number) => {
+            dispatch({
+                type: "SET",
+                page,
+            });
+        }, []),
```
Comment (erictooth) [inline src/index.ts:92]: It's true that the reference to `dispatch` should never change, but doesn't React team recommend passing these things as dependencies anyways as a good practice?

File: package.json (hunk 2 above)
Comment (atomicpages) [reply package.json]: Makes sense! I can make the change

File: src/index.ts (hunk 4 above)
Comment (atomicpages) [reply src/index.ts:92]: 🤷  not that I've seen. [From the react docs](https://reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often):

```js
function Counter() {
  const [count, setCount] = useState(0);

  useEffect(() => {
    const id = setInterval(() => {
      setCount(c => c + 1); // ✅ This doesn't depend on `count` variable outside
    }, 1000);
    return () => clearInterval(id);
  }, []); // ✅ Our effect doesn't use any variables in the component scope

  return <h1>{count}</h1>;
}
```

Happy to add them back, LMK!

File: src/index.ts (hunk 3 above)
Comment (atomicpages) [reply src/index.ts]: Good point! The current union works well, but how should `action = {}` in the reducer be initialized? With the union, `type` must be: `"SET" | "NEXT" | "PREV"`

**Pull request merged**